File: reactor/API/src/jetbrains/mps/logic/reactor/evaluation/CompositeFeedback.java
/*
 * Copyright 2014-2019 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jetbrains.mps.logic.reactor.evaluation;

import jetbrains.mps.logic.reactor.program.Rule;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Fedor Isakov
 */
public class CompositeFeedback extends EvaluationFeedback {

    /**
     * Composition is associative but not commutative.
     */
    public static EvaluationFeedback of(EvaluationFeedback left, EvaluationFeedback right) {
        if (left == null || right == null) {
            return left != null ? left : right;

        } else if (left instanceof CompositeFeedback) {
            return ((CompositeFeedback) left).composeRight(right);

        } else if (right instanceof CompositeFeedback) {
            return ((CompositeFeedback) right).composeLeft(left);

        } else {
            return new CompositeFeedback(Arrays.asList(left, right));
        }
    }

    private CompositeFeedback(List<EvaluationFeedback> newElements) {
        this.severity = maxSeverity(newElements);
        this.elements = newElements;
    }

    @Override
    public boolean handle(Rule rule, EvaluationFeedbackHandler handler) {
        int unhandled = 0;
        for (EvaluationFeedback feedback : elements) {
            if (!feedback.alreadyHandled()) {
                unhandled += 1;
                if (handler.handleFeedback(rule, feedback)) {
                    feedback.setHandled();
                    unhandled -= 1;
                }
            }
        }
        return unhandled == 0;
    }

    @Override
    public Severity getSeverity() {
        return severity;
    }

    @Override
    public String getMessage() {
        return composeMessages(elements);
    }

    public List<EvaluationFeedback> elements() {
        return elements;
    }

    @Override
    public String toString() {
        return elements.toString();
    }

    private CompositeFeedback composeLeft(EvaluationFeedback left) {
        ArrayList<EvaluationFeedback> newElements = new ArrayList<>();
        if (left instanceof CompositeFeedback) {
            newElements.addAll(((CompositeFeedback) left).elements);

        } else {
            newElements.add(left);
        }
        newElements.addAll(elements);
        return new CompositeFeedback(newElements);
    }

    private CompositeFeedback composeRight(EvaluationFeedback right) {
        ArrayList<EvaluationFeedback> newElements = new ArrayList<>(elements);
        if (right instanceof CompositeFeedback) {
            newElements.addAll(((CompositeFeedback) right).elements);

        } else {
            newElements.add(right);
        }
        return new CompositeFeedback(newElements);
    }

    private static Severity maxSeverity(List<EvaluationFeedback> efs) {
        Severity sev = Severity.DEBUG;
        for(EvaluationFeedback ef : efs) {
            if (ef.getSeverity().compareTo(sev) > 0) {
                sev = ef.getSeverity();
            }
        }
        return sev;
    }

    private static String composeMessages(List<EvaluationFeedback> efs) {
        StringBuilder sb = new StringBuilder();
        String sep = "";
        for(EvaluationFeedback ef : efs) {
            sb.append(sep).append(ef.getMessage());
            sep = "; ";
        }
        return sb.toString();
    }

    private final List<EvaluationFeedback> elements;
    private final Severity severity;
}

File: reactor/Core/src/jetbrains/mps/logic/reactor/core/CompositeFeedback.kt
/*
 * Copyright 2014-2019 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jetbrains.mps.logic.reactor.core

import jetbrains.mps.logic.reactor.evaluation.RuleMatch
import jetbrains.mps.logic.reactor.evaluation.Supervisor
import jetbrains.mps.logic.reactor.program.Rule

import java.util.ArrayList

/**
 * @author Fedor Isakov
 */
class CompositeFeedback private constructor(private val elements: List<Feedback>) : Feedback() {
    private val severity: Severity = maxSeverity(elements)

    override fun handle(supervisor: Supervisor, ruleMatch: RuleMatch, provenance: List<Rule>): Boolean {
        var unhandled = false
        if (!alreadyHandled()) {
            for (feedback in elements) {
                if (!feedback.handle(supervisor, ruleMatch, provenance)) {
                    unhandled = true
                }
            }
        }
        if (!unhandled) {
            setHandled()
        }
        return alreadyHandled()
    }

    override fun report(supervisor: Supervisor, ruleMatch: RuleMatch, provenance: List<Any>) {
        if (!alreadyHandled()) {
            for (feedback in elements) {
                feedback.report(supervisor, ruleMatch, provenance)
            }
            setHandled()
        }
    }

    override fun getSeverity(): Severity {
        return severity
    }

    override fun getMessage(): String {
        return composeMessages(elements)
    }

    override fun getDetails(): Any? {
        // FIXME compose the details
        return null
    }

    fun elements(): List<Feedback> {
        return elements
    }

    override fun toString(): String {
        return elements.toString()
    }

    private fun composeLeft(left: Feedback): CompositeFeedback {
        val newElements = ArrayList<Feedback>()
        if (left is CompositeFeedback) {
            newElements.addAll(left.elements)

        } else {
            newElements.add(left)
        }
        newElements.addAll(elements)
        return CompositeFeedback(newElements)
    }

    private fun composeRight(right: Feedback): CompositeFeedback {
        val newElements = ArrayList(elements)
        if (right is CompositeFeedback) {
            newElements.addAll(right.elements)

        } else {
            newElements.add(right)
        }
        return CompositeFeedback(newElements)
    }

    companion object {

        /**
         * Composition is associative but not commutative.
         */
        fun of(left: Feedback?, right: Feedback?): Feedback? {
            return if (left == null || right == null) {
                left ?: right

            } else (left as? CompositeFeedback)?.composeRight(right)
                ?: ((right as? CompositeFeedback)?.composeLeft(left) ?: CompositeFeedback(listOf(left, right)))
        }

        fun dropLast(maybeComposite: Feedback?): Feedback? =
            if (maybeComposite == null || maybeComposite !is CompositeFeedback) {
                null
            } else if (maybeComposite.elements.size > 2) {
                CompositeFeedback(maybeComposite.elements.dropLast(1))
            } else if (maybeComposite.elements.size == 2) {
                maybeComposite.elements[0]
            } else throw NoSuchElementException()


        private fun maxSeverity(efs: List<Feedback>): Severity {
            var sev: Severity = Severity.DEBUG
            for (ef in efs) {
                if (ef.severity > sev) {
                    sev = ef.severity
                }
            }
            return sev
        }

        private fun composeMessages(efs: List<Feedback>): String {
            val sb = StringBuilder()
            var sep = ""
            for (ef in efs) {
                sb.append(sep).append(ef.message)
                sep = "; "
            }
            return sb.toString()
        }
    }
}
